### PR TITLE
Improve error message for illegal enum variants

### LINKED ISSUE #4942: Add helpful error message when detecting Rust style enums
Users moving from Rust might be tempted to use the Rust enum syntax:

```rust
enum A {
    Foo,
    Bar(u32),
}
```

When the correct Sway syntax is: 

```sway
enum A {
    Foo: (),
    Bar: u32,
}
```

This is surprising, so we should have an error message that detects this pattern and asks the developer if they meant the latter.

### PR DESCRIPTION
**Pull request by jjcnn**
## Description

Fixes #4942. 

Please do review this carefully. I've added a test triggers the new error message, but I might very well have forgotten to add something somewhere.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: sway-error/src/parser_error.rs [hunk 1]
```diff
@@ -94,6 +94,8 @@ pub enum ParseErrorKind {
     UnexpectedTokenAfterSliceType,
     #[error("Expected a path type.")]
     ExpectedPathType,
+    #[error("Expected ':'. Enum variants must be on the form 'Foo : ()' or 'Bar : u32'.")]
```
Comment (ironcev) [inline sway-error/src/parser_error.rs:97]:
```suggestion
    #[error("Expected ':'. Enum variants must be in the form `Variant: ()`, `Variant: <type>`, or `Variant: (type1, ..., typeN)`")]
```

Comment (ironcev) [reply sway-error/src/parser_error.rs:97]: Let's do it like Rust does it, by listing all the possible forms and using the generic name "Variant". Note that according to the diagnostics RFC we enclose code in ``.

File: test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/src/main.sw [hunk 2]
```diff
@@ -0,0 +1,19 @@
+script;
+
+// A developer familiar with Rust might be tempted to define enum variants as 'Foo' or 'Bar(u32).
```
Comment (ironcev) [inline test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/src/main.sw:3]:
```suggestion
// A developer familiar with Rust might be tempted to define enum variants as `Foo` or `Bar(u32)`.
```

File: test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/src/main.sw [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  Ok,    // Illegal
+  Err,   // Also illegal, but shadowed by previous error
+}
+
+enum Enum2 {
+  F(u32),        // Illegal
+  G(u32, u32),   // Also illegal, but shadowed by previous error
+}
+
+fn main() -> () {
```
Comment (ironcev) [inline test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/src/main.sw:16]:
```suggestion
fn main() {
```

File: test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/src/main.sw [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    ()
```
Comment (ironcev) [inline test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/src/main.sw:17]:
```suggestion
```

**Review by ironcev — CHANGES_REQUESTED** [review]
A few slight changes in the message text, to make it compatible with the diagnostic guidelines.

~~Since this is a new diagnostic, we also need to add the `Diagnostic` to the error, this is missing currently.~~

We don't have `Diagnostic` so far in the parsing phase, so let's leave that part out for now.

File: test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml [hunk 5]
```diff
@@ -3,9 +3,9 @@ category = "fail"
 # check: $()error
 # check: enum_rust_like/src/main.sw:7:5
 # check: $()Ok,    // Illegal
-# nextln: ^ Expected ':'. Enum variants must be on the form 'Foo : ()' or 'Bar : u32'.
+# nextln: $()^ Expected ':'. Enum variants must be on the form 'Foo : ()' or 'Bar : u32'.
```
Comment (ironcev) [inline test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml:6]: There is no need to write the leading `^ `.

File: test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml [hunk 6]
```diff
@@ -3,9 +3,9 @@ category = "fail"
...
 
 # check: $()error
 # check: enum_rust_like/src/main.sw:12:5
 # check: $()F(u32),        // Illegal
-# nextln: ^ Expected ':'. Enum variants must be on the form 'Foo : ()' or 'Bar : u32'.
+# nextln: $()^^^^^ Expected ':'. Enum variants must be on the form 'Foo : ()' or 'Bar : u32'.
```
Comment (ironcev) [inline test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml:11]: Same here. Just `$()Expected ...` Makes tests more less prone to changes later on.

**Comment by ironcev** [comment]
@jjcnn @IGI-111 This minimal change is already a big improvement compared to what we have now. It clearly indicates the issue and the expected variant formats. So I'm definitely for merging it.

However, after giving it additional thought, I'm thinking again about a bit more ambitious solution, which was my original thought on the GitHub issue. Recognizing Rust specific patterns and showing concrete solution for the issue, how the same code should look in Sway.

That would mean adding `Diagnostic` to the `CompileError::Parse` which we don't have now, and extended the parser to actually parses two typical Rust examples. Both changes are rather simple and will allow us to have messages like these:

```
error: Enum variant declaration is not valid.

enum Enum1 {
  Ok,    // Illegal
  ^^ `Ok` is not a valid enum variant declaration.
  -- help: Did you mean `Ok: ()`?
  -- help: In Sway enum variants are in the form `Variant: ()`, `Variant: <type>`, or `Variant: (<type1>, ..., <typeN>)`.
  -- help: E.g., `Foo: (), or `Bar: (bool, u32)`.  
```

```
error: Enum variant declaration is not valid.

enum Enum2 {
  F(u32),        // Illegal
  ^^^^^^ `F(u32)` is not a valid enum variant declaration.
  ------ help: Did you mean `F: (u32)`?
  ------ help: In Sway enum variants are in the form `Variant: ()`, `Variant: <type>`, or `Variant: (<type1>, ..., <typeN>)`.
  ------ help: E.g., `Foo: (), or `Bar: (bool, u32)`.  
```

```
error: Enum variant declaration is not valid.

enum Enum3 {
  A(MyType, bool, u8),
  ^^^^^^^^^^^^^^^^^^^ `A(MyType, bool, u8)` is not a valid enum variant declaration.
  ------------------- help: Did you mean `A: (MyType, bool, u8)`?
  ------------------- help: In Sway enum variants are in the form `Variant: ()`, `Variant: <type>`, or `Variant: (<type1>, ..., <typeN>)`.   
  ------------------- help: E.g., `Foo: (), or `Bar: (bool, u32)`.  
```

We do not have to parse the Rust struct-like variant. That would be a bit too much work, and it's also not a kind of a mistake Sway programmers do. In that and all other cases we can go for the common help message without the particular suggestion.

```
error: Enum variant declaration is not valid.

enum Enum4 {
  A { x: i32, y: i32 },
  ^^^^^^^^^^^^^^^^^^^ `A { x: i32, y: i32 }` is not a valid enum variant declaration.
  ------------------- help: In Sway enum variants are in the form `Variant: ()`, `Variant: <type>`, or `Variant: (<type1>, ..., <typeN>)`.
  ------------------- help: E.g., `Foo: (), or `Bar: (bool, u32)`.  
```

But for the other two cases, having a precise solution in the error message has a benefit.

Also, adding the first `Diagnostic` for `CompileError::Parse` will pay off in the mid and long term.

Let me know what you think.

**Comment by IGI-111** [comment]
I think it is a good idea to get additional information to help the most common issues, but unless Rust confusions are really that common it's probably over-engineering.

Helpful errors are important and I really like your examples, but improving errors for cases that aren't as niche is likely a better use of our ressources at this time. Maybe you should put this idea in a separate issue and we can bikeshed it and/or pick it up when we'll have a serious go at polishing our errors.

In any case, we should get this PR's improvement first.

**Comment by ironcev** [comment]
> Maybe you should put this idea in a separate issue and we can bikeshed it and/or pick it up when we'll have a serious go at polishing our errors.

Done: #5403

File: test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml (hunk 5 above)
Comment (jjcnn) [reply test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml]: Fixed.

File: test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml (hunk 6 above)
Comment (jjcnn) [reply test/src/e2e_vm_tests/test_programs/should_fail/enum_rust_like/test.toml]: Fixed.

**Comment by jjcnn** [comment]
Review comments addressed.

**Review by IGI-111 — APPROVED** [review]

**Review by ironcev — APPROVED** [review]

**Pull request merged**